feat(auto-instrumentations-node): enabling instrumentations via env

atsu85 opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

Allowing to speed up application startup time by allowing to disable instrumentations that are not needed via environment variable for use-cases where it isn't possible to disable instrumentations via `inputConfigs` object passed to [getNodeAutoInstrumentations(inputConfigs)](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/d0fc194ace159bbcd56cf9c452348ab148fd09c4/metapackages/auto-instrumentations-node/src/utils.ts#L135) (like when using Kubernetes opentelemetry-operator - its [instrumentation code](https://github.com/open-telemetry/opentelemetry-operator/blob/d9e1d161366dbf967a9cb47f3d6ee6137ec4d8d6/autoinstrumentation/nodejs/src/autoinstrumentation.ts#L37) can't be customized by applications running in the Kubernetes cluster). More information in [this comment](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1672#issuecomment-1956839326)

## Short description of the changes

This PR allows enabling  instrumentations selected via `OTEL_NODE_INSTRUMENTATIONS` environment variable while disabling others. 

To keep backwards compatibility, by default all instrumentations are still enabled when the environment variable isn't provided.

## Usage example:

as documented in the [README change](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/3c4bc0aa1ccca94280cd0e2dc562913812577357#diff-01f7f78cee341bacd34c71334c45b2d8299aa4d66c6b7951950389f22f625cfe):

To enable only
[@opentelemetry/instrumentation-http](https://github.com/open-telemetry/opentelemetry-js/tree/main/packages/opentelemetry-instrumentation-http)
and [@opentelemetry/instrumentation-nestjs-core](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-nestjs-core)
instrumentations:

```shell
export OTEL_NODE_INSTRUMENTATIONS="http,nestjs-core"
```

## Alternative approaches considered

In [this comment](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1672#issuecomment-1956839326) I proposed two approaches how to enable/disable specific instrumentations via environment variables.

This PR implements the first approach. The benefits of this approach are:
* it allows disabling all non-required instrumentations with single environment variable (easier to configure, similar to [configuring resource detectors via OTEL_NODE_RESOURCE_DETECTORS](https://opentelemetry.io/docs/languages/js/automatic/configuration/#sdk-resource-detector-configuration))
* when new instrumentations are added here ([@opentelemetry/auto-instrumentations-node](https://www.npmjs.com/package/@opentelemetry/auto-instrumentations-node) npm package), they won't be enabled by default when only selected instrumentations were enabled, and hence won't degrade application startup time (no need to add additional environment variables to disable each new instrumentation when this package is updated)

This does not prevent implementing later also second approach backwards compatibly.
Could implement it for example so that globally only few instrumentations are enabled for all applications using `OTEL_NODE_INSTRUMENTATIONS` and each application could enable additional instrumentations with second approach based on additional environment variables for each instrumentation

atsu85 commented:
I thought I'll share the investigation results that motivated this PR:

# Summary

## Auto-instrumentation is much slower compared to instrumentation after application has loaded dependencies:

> I tested this in our Kubernetes cluster

When performing instrumentation after loading application bootstrap import modules, application startup took just 1.5 seconds. Auto-instrumentation (with `--require path/to/instrumentation.js`) is a lot slower when all instrumentations are enabled:

* ca 27 times slower in case of quite low CPU limit (that throttles CPU a lot: 300m)

* ca 6.5 times slower in case of quite high CPU limit (that prevents CPU throttling: 1500m)

## Disabling resource detectors didn’t have a visible effect on startup time with instrumentation agent

## In which startup phase the time is spent?

> I tested this part locally on much more resources than allocated for the application in our Kubernetes cluster

* The slowest startup phases are importing modules for bootstraping the application and starting NestJS:

    * loading auto-instrumentation: 0 vs 374ms

    * importing modules (after loading instrumentation): 244 vs 4033ms

    * starting app (after importing modules): 123 vs 2714ms

* disabling all instrumentation besides http and nest-js will add just 0.4s startup overhead on my machine

# Profiling application startup locally revealed that almost all the time is wasted on loading node modules

* because instrumentation is done before loading application bootstrap modules (probably gazillions of NodeJS modules and each module is instrumented by [37 instrumentations](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/038e0bfda951055ce91724a3b4a3042a9f918700/metapackages/auto-instrumentations-node/src/utils.ts#L87-L124) enabled by default) 
![image](https://github.com/open-telemetry/opentelemetry-js-contrib/assets/1206512/a65e2bf5-1fd5-4de3-bf15-43743b9adc0c)


    * when instrumentation was done after loading application bootstrap modules (manually in code, not via --require ...), the overhead for the startup was insignificant

* When i disabled all instrumentations, then application started up almost without any overhead to the application startup process (besides mainly loading the agent - ca 0.4 seconds)

atsu85 commented:
@obecny, could you take a look?

obecny commented on metapackages/auto-instrumentations-node/src/utils.ts:
```diff
@@ -167,6 +166,34 @@ export function getNodeAutoInstrumentations(
+}
+
+/**
+ * Returns the list of instrumentations that are allowed based on the environment variable.
```
```suggestion
 * Returns the list of instrumentations that are enabled based on the environment variable.
```

atsu85 replied:
done with [this commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/5f7a4cf6698b210bbc59e65fcda6e4c05e73b804), that is squashed into [the feature commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/c7393d9ec7dddbdc7031d37a420d69fd30e11614) after rebasing on top of latest main

obecny commented on metapackages/auto-instrumentations-node/src/utils.ts:
```diff
@@ -135,12 +135,8 @@ export type InstrumentationConfigMap = {
-    }
-  }
+  checkManuallyProvidedInstrumentationNames(Object.keys(inputConfigs));
+  const instrumentationsFromEnv = getAllowedInstrumentationsFromEnv();
```
```suggestion
  const enabledInstrumentationsFromEnv = getEnabledInstrumentationsFromEnv();
```

atsu85 replied:
renamed with [this commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/5f7a4cf6698b210bbc59e65fcda6e4c05e73b804), that is squashed into [the feature commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/c7393d9ec7dddbdc7031d37a420d69fd30e11614) after rebasing on top of latest main

obecny reviewed (commented):
thx for this contribution, few suggestions with regards to naming so they are more descriptive, other than that would lgtm, thx

atsu85 commented on metapackages/auto-instrumentations-node/README.md:
````diff
@@ -75,6 +75,21 @@ For example, to enable only the `env`, `host` detectors:
+instrumentations:
+
+```shell
+export OTEL_NODE_INSTRUMENTATIONS="http,nestjs-core"
````
Thanks for the comment. I even considered that naming convention, but initially decided for the one i used to align with existing env variable name `OTEL_NODE_RESOURCE_DETECTORS` used in the same project. 

@obecny, in that light would you still suggest using the name you proposed? I don't mind renaming it, just let me know

atsu85 replied:
oh, renaming makes even more sense after i read [this comment](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953#issuecomment-1966088288)

atsu85 replied:
renamed with [this commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/3f547dd51e2a01ca7e53e2eb9818e6c12c96ffe5), that is squashed into [the feature commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/c7393d9ec7dddbdc7031d37a420d69fd30e11614) after rebasing on top of latest main

atsu85 commented on metapackages/auto-instrumentations-node/src/utils.ts:
```diff
@@ -167,6 +166,34 @@ export function getNodeAutoInstrumentations(
+/**
+ * Returns the list of instrumentations that are allowed based on the environment variable.
+ */
+function getAllowedInstrumentationsFromEnv() {
```
renamed with [this commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/5f7a4cf6698b210bbc59e65fcda6e4c05e73b804), that is squashed into [the feature commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/c7393d9ec7dddbdc7031d37a420d69fd30e11614) after rebasing on top of latest main

atsu85 commented:
@obecny, thanks for super-quick feedback.

[Adding another variable OTEL_DISABLED_NODE_INSTRUMENTATIONS](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953#issuecomment-1966088288) sounds good, but i would leave it out of this PR, because 

* I'm not sure if after this feature people even would start using it. I'm dogfooding content currently present in this PR, but i wouldn't dogfood the OTEL_DISABLED_NODE_INSTRUMENTATIONS part, and i'm not sure if anyone else would start using it either.
* implementing it would be fairly easy without introducing breaking changes related to this PR

What do you think?

obecny commented:
> @obecny, thanks for super-quick feedback.
> 
> [Adding another variable OTEL_DISABLED_NODE_INSTRUMENTATIONS](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953#issuecomment-1966088288) sounds good, but i would leave it out of this PR, because
> 
> * I'm not sure if after this feature people even would start using it. I'm dogfooding content currently present in this PR, but i wouldn't dogfood the OTEL_DISABLED_NODE_INSTRUMENTATIONS part, and i'm not sure if anyone else would start using it either.
> * implementing it would be fairly easy without introducing breaking changes related to this PR
> 
> What do you think?

definitely in next PR, that was just suggestions and some more info why having "enabled" in name would make possible to have 2 features and then user can set it much easier.

obecny reviewed (approved):
thx for changes, lgtm

atsu85 commented:
@obecny, thanks for the feedback and approving the PR!

However 2 topics:

1) Perhaps You'll mark CR comment threads also as resolved? (I could do it myself, but IMHO the reviewer should decide if it is resolved or not ;) ) 

2) GH is still showing that "Merging is blocked" and "Code owner review required":
![image](https://github.com/open-telemetry/opentelemetry-js-contrib/assets/1206512/2da3df16-be2e-4394-8c60-1fce3c668b99)
most likely because of [this line in CODEOWNERS file](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/.github/CODEOWNERS#L15),
but that team either doesn't exist ([this link](https://github.com/orgs/open-telemetry/teams/javascript-approvers) doesn't work for me) or it exists and I just can't see it because I'm not a member of that team?

How to proceed with merging it?

atsu85 commented:
@obecny, can you merge the PR yourself or should i do smth else myself to get it merged?
After reading again [Pull Request Merge Requirements](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/CONTRIBUTING.md#pull-request-merge-requirements) from contribution guide, it seems that you should be able to merge it (I also updated my [previous comment](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953#issuecomment-1966455951) regarding 2 topics I raised yesterday)

atsu85 commented:
[@open-telemetry/javascript-approvers](https://github.com/orgs/open-telemetry/teams/javascript-approvers), please approve this PR - @obecny who is the [component owner for auto-instrumentations-node](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/.github/component_owners.yml#L20-L21) has already aproved it:

atsu85 commented:
@blumamir, can you merge this PR? [Looks like you have the merge permissions](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1772#event-10899892092) that the component-owner (who approved this PR doesn't have)

pichlermarc reviewed (commented):
Sorry, was busy with other PRs and was not able to pick up a new feature to review yet.
Also it looks like this package was not owned by the maintainers so we did not get auto-assigned (opened #1996 to fix that)

Generally looks good % comments (env var naming and lint fixes) :+1:

pichlermarc commented:
> I'm not sure if after this feature people even would start using it. I'm dogfooding content currently present in this PR, but i wouldn't dogfood the OTEL_DISABLED_NODE_INSTRUMENTATIONS part, and i'm not sure if anyone else would start using it either.

I agree with this. I think it's good to hold off on implementing this and would argue that a follow up to implement this is not necessary. Let's see if there's people who have a real need for it first. 

I think the here implemented option of explicitly opting-in to instrumentations will be preferred by most.

atsu85 commented on metapackages/auto-instrumentations-node/README.md:
````diff
@@ -75,6 +75,21 @@ For example, to enable only the `env`, `host` detectors:
 export OTEL_NODE_RESOURCE_DETECTORS="env,host"
 ```
 
+
````
removed with [this commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/74cb0d725d95e198df2ef0f9b9fc0d0d4f1f1da1), that is squashed into [the feature commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/d257d027abd2ffd5cd54186f218cc5a24357997f) after rebasing on top of latest main

atsu85 commented on metapackages/auto-instrumentations-node/README.md:
````diff
@@ -75,6 +75,21 @@ For example, to enable only the `env`, `host` detectors:
+export OTEL_ENABLED_NODE_INSTRUMENTATIONS="http,nestjs-core"
+```
+
+
````
removed with [this commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/74cb0d725d95e198df2ef0f9b9fc0d0d4f1f1da1), that is squashed into [the feature commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/d257d027abd2ffd5cd54186f218cc5a24357997f) after rebasing on top of latest main

atsu85 commented on metapackages/auto-instrumentations-node/README.md:
````diff
@@ -75,6 +75,21 @@ For example, to enable only the `env`, `host` detectors:
+instrumentations:
+
+```shell
+export OTEL_ENABLED_NODE_INSTRUMENTATIONS="http,nestjs-core"
````
renamed with [this commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/74cb0d725d95e198df2ef0f9b9fc0d0d4f1f1da1), that is squashed into [the feature commit](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1953/commits/d257d027abd2ffd5cd54186f218cc5a24357997f) after rebasing on top of latest main

atsu85 commented:
@pichlermarc, thanks for the feedback. 
I've addressed all the comments, so please review again (and mark them as resolved yourself after verifying them).

atsu85 commented:
@pichlermarc, thanks for approving! 
Can you merge and release it as well, or is there anything I could do myself, without having merge permissions?

pichlermarc commented:
> @pichlermarc, thanks for approving! Can you merge and release it as well, or is there anything I could do myself, without having merge permissions?

I'm going to merge it now, but the release will only happen next week as I'd like to avoid releasing on a Friday.

hamin commented:
Hurray! Was waiting for this for so long!
